util/winutil/restartmgr_windows.go

```diff
@@ -0,0 +1,830 @@
+	if err != nil {
+		return nil, fmt.Errorf("ProcessImageName: %w", err)
+	}
+
+	const _RESTART_MAX_CMD_LINE = 1024
+	var cmdLine [_RESTART_MAX_CMD_LINE]uint16
+	cmdLineLen := uint32(len(cmdLine))
+	var rmFlags uint32
+	hr := getApplicationRestartSettings(h, &cmdLine[0], &cmdLineLen, &rmFlags)
+	// Not found is not an error; it just means that the app never set any restart settings.
+	if e := wingoes.ErrorFromHRESULT(hr); e.Failed() && e != wingoes.ErrorFromErrno(windows.ERROR_NOT_FOUND) {
+		return nil, fmt.Errorf("GetApplicationRestartSettings: %w", error(e))
+	}
+	if (rmFlags & _RESTART_NO_PATCH) != 0 {
+		// The application explicitly stated that it cannot be restarted during
+		// an upgrade.
+		return nil, ErrProcessNotRestartable
+	}
+
+	// These token bits are non-fatal, so we'll proceed with best-effort.
+	var userSID string
+	var tok windows.Token
+	if err := windows.OpenProcessToken(h, windows.TOKEN_QUERY, &tok); err == nil {
+		defer tok.Close()
+		if tokenUser, err := tok.GetTokenUser(); err == nil {
+			// Save the user's SID so that we can later check it against the currently
+			// logged-in Tailscale profile.
+			userSID = tokenUser.User.Sid.String()
+		}
+	}
+
+	result := &RestartableProcess{
+		Process: *up,
+		SessionInfo: SessionID{
+			LogonSession: logonSessionID,
+			TSSession:    tsSessionID,
+		},
+		CommandLineInfo: CommandLineInfo{
+			ExePath: img,
+			Args:    windows.UTF16ToString(cmdLine[:cmdLineLen]),
+		},
+		UserSID: userSID,
+		handle:  h,
+	}
+
+	runtime.SetFinalizer(result, func(rp *RestartableProcess) { rp.Close() })
+	h = 0
+	return result, nil
+}
+
+// RestartableProcess contains the necessary information to uniquely identify
+// an existing process, as well as the necessary information to be able to
+// terminate it and later start a new instance in the identical logon session
+// to the previous instance.
+type RestartableProcess struct {
+	// Process uniquely identifies the existing process.
+	Process UniqueProcess
+	// SessionInfo uniquely identifies the Terminal Services (RDP) and logon
+	// sessions the existing process is running under.
+	SessionInfo SessionID
+	// CommandLineInfo contains the command line information necessary for restarting.
+	CommandLineInfo CommandLineInfo
+	// UserSID contains the stringified SID of the existing process's token's user.
+	UserSID string
+	// handle specifies the Win32 HANDLE associated with the existing process.
+	// When non-zero, it includes access rights for querying, terminating, and synchronizing.
+	handle windows.Handle
+	// hasExitCode is true when the exitCode field is valid.
+	hasExitCode bool
+	// exitCode contains exit code returned by this RestartableProcess once
+	// its termination has been recorded by (RestartableProcesses).Terminate.
+	// It is only valid when hasExitCode == true.
+	exitCode uint32
+}
+
+func (rp *RestartableProcess) Close() error {
+	if rp.handle == 0 {
+		return nil
+	}
+	if err := windows.CloseHandle(rp.handle); err != nil {
```

Comment:
If this fails, should `handle` still be reset to 0? Presumably the handle is no good if this produces an error.